### PR TITLE
feat: Create a build-time-optimized release build profile

### LINKED ISSUE #3316: Use custom cargo profiles to create a release build profile optimized more for build time
Custom cargo profiles: https://github.com/rust-lang/cargo/pull/9943/

https://github.com/influxdata/influxdb_iox/pull/3081 changed settings to optimize for binary size and runtime performance, but it takes longer to build.

Make a custom profile that makes a release build that will perform better than a debug build, but will build faster.

### PR DESCRIPTION
**Pull request by carols10cents**
Closes #3316

A "full build" was me deleting `/target` and then running the command. A "rebuild" was me changing a particular `unreachable!` to a `panic!` in the `read_buffer` crate.

| Command | Full build | Rebuild |
|-----------|-----------|---------|
| `cargo build --release` |  7m 24s | 5m 13s |
| `cargo build --profile quick-release` | 4m 11s | 2m 06s |

Bikeshed time:

- Should the profile optimizing for compile time be `--release` or should the profile optimizing for binary size be `--release`?
- What should the name of the non-`--release` profile be?
- Any other comparison stats you want to see?
- Any other settings that should be changed?

### CODE REVIEW

**Comment by e-dard** [comment]
I like the idea of `release` and `quick-release`! This makes me happy and is just what I was looking for!

**Review by e-dard — APPROVED** [review]

File: Cargo.toml [hunk 1]
```diff
@@ -85,3 +85,8 @@ lto = "thin"
 
 [profile.bench]
 debug = true
+
+[profile.quick-release]
```
Comment (alamb) [inline Cargo.toml:89]: I suggest some description of the difference of `release` and `quick-release`

Perhaps something like that `release` takes significantly longer to compile but may make a faster binary, and that `quick-release` is for dev iterations

**Review by alamb — APPROVED** [review]
Thank you @carols10cents

File: Cargo.toml (hunk 1 above)
Comment (carols10cents) [reply Cargo.toml:89]: Good call! Just added!

File: Dockerfile [hunk 2]
```diff
@@ -14,14 +14,12 @@ COPY . /influxdb_iox
 WORKDIR /influxdb_iox
 
 ARG CARGO_INCREMENTAL=yes
-ARG CARGO_PROFILE_RELEASE_CODEGEN_UNITS=1
-ARG CARGO_PROFILE_RELEASE_LTO=thin
+ARG PROFILE=release
```
Comment (carols10cents) [inline Dockerfile:17]: @crepererum is this what you wanted to happen in the Dockerfile with regards to this change? I'm not sure that I've understood correctly. Thanks!

Comment (crepererum) [reply Dockerfile:17]: This is precisely what I wanted, thanks! :mrs_claus:

**Review by crepererum — APPROVED** [review]

**Review by crepererum — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 3]
```diff
@@ -31,7 +29,7 @@ RUN \
   --mount=type=cache,id=influxdb_iox_git,sharing=locked,target=/usr/local/cargo/git \
   --mount=type=cache,id=influxdb_iox_target,sharing=locked,target=/influxdb_iox/target \
     du -cshx /usr/local/cargo/registry /usr/local/cargo/git /influxdb_iox/target && \
-    cargo build --target-dir /influxdb_iox/target --release --no-default-features --features="$FEATURES" && \
+    cargo build --target-dir /influxdb_iox/target --profile="$PROFILE" --no-default-features --features="$FEATURES" && \
     objcopy --compress-debug-sections target/release/influxdb_iox && \
     cp /influxdb_iox/target/release/influxdb_iox /root/influxdb_iox && \
     du -cshx /usr/local/cargo/registry /usr/local/cargo/git /influxdb_iox/target
```
Comment (crepererum) [inline Dockerfile:35]:
```suggestion
    objcopy --compress-debug-sections "target/$PROFILE/influxdb_iox" && \
    cp "/influxdb_iox/target/$PROFILE/influxdb_iox" /root/influxdb_iox && \
    du -cshx /usr/local/cargo/registry /usr/local/cargo/git /influxdb_iox/target
```

Comment (carols10cents) [reply Dockerfile:35]: Oooooooh yep. That'll be a problem. Great catch!

**Comment by carols10cents** [comment]
@crepererum I think this is ready for re-review, thank you!

**Review by crepererum — APPROVED** [review]
:tada:

**Pull request merged**